# **Variable and Data Types**

Numbers

In [ ]:
varNum = 123
pi = 3.14159

String

In [ ]:
varString = "Hello World"
varText = 'This is a String'

List


In [ ]:
varList = ["abc", 123]

Tuples

In [ ]:
varTuple = 'abc', 123, "HELLO"

Dictionaries

In [ ]:
var = 3
varDict = {'first':1 , '2' : '2nd' , 3:var}

In [ ]:
varDict = {}
varDict['first'] = 1
varDict['2'] = '2nd'
varDict[3] = var

# **Arithmetic**


Addition

In [ ]:
a = 5 + 3

>8


Subtraction

In [ ]:
a = 5 - 3

>2

Multiplication

In [ ]:
a = 5 * 3

>15

Exponent

In [ ]:
a = 5 ** 3

>125

Division

In [ ]:
a = 5 / 3

1.666666666666666666667

In [ ]:
a = 5 % 3

>2

In [ ]:
a = 5//3

>1

Increment/Decrement

In [ ]:
a = 5
a += 1

>6

Decrement

In [ ]:
a = 5
a -= 1

>4

String Concatenation

In [ ]:
a = 'Hello ' + 'World'

>Hello World!

Complex Expressions

In [ ]:
a = 3 + 5 - 6 * 2 / 4

>5.0

# **Control Statement and Data Structures**

Boolean Conditions

In [ ]:
x = True

if x:
  print("var x is True")
else:
  print("var x is False")

var x is True


>var x is True

String Conditions

In [ ]:
x ="Hello World!"

if x == 'Hello World!':
  print("var x is Hello World!")
else:
  print("var x is not Hello World!")

var x is Hello World!


>var x is Hello World!

**Numerical Conditions**

In [ ]:
x = 10

if x == '10':
  print("var x is a string")
elif x == 10:
  print("var x in an integer")
else:
  print("var x is none of the above")

var x in an integer


>var x is an integer

# **Loops**

For Loops

In [ ]:
for var in range(0, 5, 2):
  print(var)

0
2
4


>0





>2





>4

While Loops

In [ ]:
var = 0
while var < 5:
  print(var)
  var += 2

0
2
4


>0





>2





>4

Nested Loops

In [ ]:
x = 0
while x < 5:
  for y in range (0,x):
    print(y,end='')
  x+=1
  print()


0
01
012
0123


>0





>01





>012





>0123

**NOTE** : The loop declaration ends with a colon (:) while its contents inside are indented

# **List**

Can contain a number of values comprised of different datatypes

In [ ]:
pi = 3.14159
varList = [1, 2, 'A' , 'B' , 'Hello!' , pi]
print(varList[0])

1


>1

In [ ]:
print(varList[4])

Hello!


In [ ]:
varList.append('World!')
print(varList[6])

World!


In [ ]:
len(varList)

7

In [ ]:
print(varList[5])

3.14159


In [ ]:
varList.remove(pi)
print(varList[5])

World!


**NOTE** : List starts with an index of 0. The remove() methd erases the first value that matches the parameter

# **Dictionaries**

In [ ]:
var = "Hello World!"
varDict = {'first' : 123, 2 : 'abc', '3' : var, 4:['lista', 'listb']}
print(varDict['first'])

123


In [ ]:
print(varDict[2])

abc


In [ ]:
print(varDict['3'])

Hello World!


In [ ]:
print(varDict[4])

['lista', 'listb']


In [ ]:
print(varDict[4][1])

listb


In [ ]:
len(varDict)

4

# **Generators and Comprehension**

Generators can be used to build list in the memory as objects

In [ ]:
def gen_num_up_to(n):
  num = 0
  while num < n:
    yield num
    num += 1

In [ ]:
gen_num_up_to(5)

<generator object gen_num_up_to at 0x7e093137d2a0>

**NOTE** : This creates an object

In [ ]:
varList = gen_num_up_to(5)
print([var for var in varList])

[0, 1, 2, 3, 4]


In [ ]:
def gen_num_up_to(n):
  num = 0
  while num < n:
    yield num
    num += 2

varList = gen_num_up_to(5)
print([var for var in varList])

[0, 2, 4]


In [ ]:
varList = range(0, 5, 2)
print([var for var in varList])

[0, 2, 4]


**NOTE** : Python's built in range function range([start] , stop , [step])

# **Slicing**

In [ ]:
varList = [1,2,3,4,5,6,7,8,9,10]
print(varList[:5])

[1, 2, 3, 4, 5]


In [ ]:
print(varList[5:])

[6, 7, 8, 9, 10]


In [ ]:
print(varList[:-2])


[1, 2, 3, 4, 5, 6, 7, 8]


In [ ]:
print(varList[-2:])

[9, 10]


In [ ]:
|print(varList[2:-2])

[3, 4, 5, 6, 7, 8]


In [ ]:
print(varList[2:8:2])

[3, 5, 7]


**NOTE** :  
list(start:)
            
list(:end)

list(start:end)

list(start:end:step)

# **Function**

Python functions uses the following notations
  
  def *funcion_name:*
     
  commands

In [ ]:
def remainder(n , m):
  while True:
    if n - m < 0:
      return n
    else:
      n = n - m
remainder(10 , 4)


2

**NOTE** : Functions are defined with the keywords def before the function_name. Similar to loops, the function definition ends with a colon (:) while its contents inside are indented

# **Exercise 2**

**Challenge 2 : Coin Flip**

Create a function coin flip() that simulates coin flips repeated times, with 0 representing tails and 1 representing heads. Make the function a generator with the parameter as the number of coin flips.

Use the numpy library as np and the function of np.random.randint() in order to simulate a coin flip. After defining this function, return the result as a List using List Comprehension discussed previously.

In [7]:
import numpy as np

def coin_flip(n):
  for _ in range(n):
        yield np.random.randint(0, 2)

print([var for var in coin_flip(8)])

[0, 1, 0, 0, 0, 1, 0, 1]
